web/app.js

```diff
@@ -2674,18 +2729,96 @@ function webViewerWheel(evt) {
 
 function webViewerTouchStart(evt) {
   if (evt.touches.length > 1) {
-    // Disable touch-based zooming, because the entire UI bits gets zoomed and
-    // that doesn't look great. If we do want to have a good touch-based
-    // zooming experience, we need to implement smooth zoom capability (probably
-    // using a CSS transform for faster visual response, followed by async
-    // re-rendering at the final zoom level) and do gesture detection on the
-    // touchmove events to drive it. Or if we want to settle for a less good
-    // experience we can make the touchmove events drive the existing step-zoom
-    // behaviour that the ctrl+mousewheel path takes.
     evt.preventDefault();
+
+    if (evt.touches.length !== 2) {
+      PDFViewerApplication._touchInfo = null;
+      return;
+    }
+
+    const [touch0, touch1] = evt.touches;
+    PDFViewerApplication._touchInfo = {
+      centerX: (touch0.pageX + touch1.pageX) / 2,
+      centerY: (touch0.pageY + touch1.pageY) / 2,
+      distance:
+        Math.hypot(touch0.pageX - touch1.pageX, touch0.pageY - touch1.pageY) ||
+        1,
+    };
   }
 }
 
+function webViewerTouchMove(evt) {
+  if (!PDFViewerApplication._touchInfo || evt.touches.length !== 2) {
+    return;
+  }
+
+  const { pdfViewer, _touchInfo, supportsPinchToZoom } = PDFViewerApplication;
+  const [touch0, touch1] = evt.touches;
+  const { pageX: page0X, pageY: page0Y } = touch0;
+  const { pageX: page1X, pageY: page1Y } = touch1;
+  const distance = Math.hypot(page0X - page1X, page0Y - page1Y) || 1;
+  const { distance: previousDistance } = _touchInfo;
+  const scaleFactor = distance / previousDistance;
+
+  if (supportsPinchToZoom && Math.abs(scaleFactor - 1) <= 1e-2) {
+    // Scale increase/decrease isn't significant enough.
+    return;
+  }
+
+  const { centerX, centerY } = _touchInfo;
+  const diff0X = page0X - centerX;
+  const diff1X = page1X - centerX;
+  const diff0Y = page0Y - centerY;
+  const diff1Y = page1Y - centerY;
+  const dotProduct = diff0X * diff1X + diff0Y * diff1Y;
+  if (dotProduct >= 0) {
+    // The two touches go in almost the same direction.
+    return;
+  }
+
+  evt.preventDefault();
+
+  const previousScale = pdfViewer.currentScale;
+
+  if (supportsPinchToZoom) {
+    PDFViewerApplication.rescale(scaleFactor);
```

Comment:
Also, can't we change this as follows and just remove the `rescale`-method?
```suggestion
    if (scaleFactor < 1) {
      PDFViewerApplication.zoomOut(/* steps = */ null, scaleFactor);
    } else if (scaleFactor > 1) {
      PDFViewerApplication.zoomIn(/* steps = */ null, scaleFactor);
    }
```
Assuming that we also change the existing methods like so:
```javascript
zoomIn(steps, scaleFactor) {
    if (this.pdfViewer.isInPresentationMode) {
      return;
    }
    this.pdfViewer.increaseScale({
      drawingDelay: AppOptions.get("defaultZoomDelay"),
      steps,
      scaleFactor,
    });
  },

  zoomOut(steps, scaleFactor) {
    if (this.pdfViewer.isInPresentationMode) {
      return;
    }
    this.pdfViewer.decreaseScale({
      drawingDelay: AppOptions.get("defaultZoomDelay"),
      steps,
      scaleFactor,
    });
  },
```